[www] Cache best block and update using dcrdata subscription.

martonp opened the pull request:
This diff subscribes politeiawww to the newblock event when it is in www mode and caches the current best block. This cached value is used in requests where the current best block is needed, but the old method of requesting it from politeiad each time is used as a fallback when there is no active websocket connection.

This diff also adds a reconnect method to `wsDcrdata` that creates a new websocket client, and resubscribes to all of the events the old client was subscribed to. This new method is used in the case of a hang up message or a read timeout from dcrdata.

Closes issue #941.

martonp commented:
@lukebp So I'll just make a separate file for pi that just subscribes to the best block. I'll add the best block subscription to the cms file too though.

lukebp commented:
> So I'll just make a separate file for pi that just subscribes to the best block.

You don't need a separate file for this.  You just need a single function, something like `setupNewBlockSub`in `politeiawww.go`, that sets up and runs the new block websocket subscription.

Also, you have data races all over this PR that need to be fixed.  You can't access shared state without holding the lock.

>  I'll add the best block subscription to the cms file too though.

Don't. Keep this PR to pi functionality only.

martonp commented:
Still working on this, don't review yet.

martonp commented:
@lukebp This is ready to be reviewed.

chappjc commented on politeiawww/politeiawww.go:
```diff
@@ -858,6 +866,92 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+			case *exptypes.WebsocketBlock:
+				log.Debugf("Message WebsocketBlock(height=%s)", m.Block.Height)
+				p.updateBestBlock(uint64(m.Block.Height))
+			case *pstypes.HangUp:
```
Glad to see this is being handled.  The pubsub server that send hangup message is deployed now.

lukebp commented:
@martonp now that #1030 has been merged, can you get this PR up to date again?

martonp commented:
@lukebp Done

martonp commented on politeiawww/politeiawww.go:
```diff
@@ -858,6 +866,92 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+				break
+			}
+			if msg == nil {
+				log.Errorf("ReceiveMsg failed")
```
I don't think it would be.. I just have it in there for safety.

martonp commented:
@lukebp This is ready to review again. The issues you raised were mostly ones which I copied from the CMS websocket listener, so we should probably update that one as well.

martonp commented on politeiawww/wsdcrdata.go:
```diff
@@ -90,9 +100,73 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+				if err == nil {
+					w.client = c
+					createdNewClient = true
+					continueListening <- true
```
I'm going to move this to after all the subscriptions are done. I'll make that change tonight.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -858,6 +865,60 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+
+			switch m := msg.Message.(type) {
+			case *exptypes.WebsocketBlock:
+				log.Debugf("Message WebsocketBlock(height=%s)", m.Block.Height)
```
```suggestion
				log.Debugf("wsDcrdata message WebsocketBlock(height=%s)",
					m.Block.Height)
```

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -90,9 +100,73 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+		timeToWait := 1 * time.Minute
+		createdNewClient := false
+
+		for {
```
What happens if the client gets hung up on during the re-connection process?

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -90,9 +100,73 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+	w.client.Stop()
+	w.removeAllSubs()
+
+	go func() {
```
What happens if the connection goes bad again during this process?

lukebp replied:
A couple points about this function.

1. What happens if the connection goes bad again during this process?
2. In general, you want to minimize nested code whenever possible.  Stuff like not using an `} else {` clause if you don't have to.  It keeps the code cleaner and easier to read.
3. Try to stay away from for loops without an ending condition.  It leads to bugs.

Here's a code snippet with some suggestions.  I haven't tested it at all, but this is the way I would write something like this.

```
func (w *wsDcrdata) reconnect() {
	done := make(chan struct{})

	go func() {
		prevSubscriptions := make(map[string]struct{}, len(w.subscriptions))
		timeToWait := 1 * time.Minute

		for len(prevSubscriptions) > 0 {
			// Reconnect to dcrdata if needed
			err := w.client.Ping()
			if err != nil {
				w.client.Stop()

				// Existing subscriptions have gone bad
				for sub := range w.subscriptions {
					prevSubscriptions[sub] = struct{}{}
				}
				w.removeAllSubs()

				// Reconnect
				c, err := initDcrdataWSClient()
				if err != nil {
					goto wait
				}
				w.client = c
			}

			// Resubscribe
			for sub := range prevSubscriptions {
				if w.isSubscribed(sub) {
					delete(prevSubscriptions, sub)
					continue
				}
				_, err := w.client.Subscribe(sub)
				if err != nil {
					log.Errorf("wsDcrdata failed to subscribe to %v: %v", err)
					goto wait
				}
				w.addSub(sub)
				delete(prevSubscriptions, sub)
			}

		wait:
			log.Debugf("wsDcrdata reconnect waiting %v", timeToWait)
			time.Sleep(timeToWait)
			timeToWait = 2 * timeToWait
		}

		// All previous subscriptions have been re-subscribed
		// to. We're done.
		done <- struct{}{}
	}()

	<-done
}
```

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -90,9 +100,73 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+}
+
+// reconnect stops the current client, initializes a new one, and subscribes
+// to each of the subscriptions that the previous client was subscribed to.
```
If you're passing in a channel the comment needs to explain what its being used for and when messages are emitted to it.

martonp commented on politeiawww/wsdcrdata.go:
```diff
@@ -90,9 +100,73 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+	w.client.Stop()
+	w.removeAllSubs()
+
+	go func() {
```
Yeah, this is way better, thanks. I'll write more like this from now on.

martonp commented:
@lukebp It's ready to be reviewed again.

martonp commented:
@lukebp CI wasn't able to install the linters. It worked a minute before and all I updated was a comment. I'm not able to rerun it since I don't have write access.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -858,6 +865,60 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+				log.Infof("Dcrdata has hung up. Will reconnect...")
+				p.wsDcrdata.reconnect()
+			default:
+				log.Errorf("Message of type %v unhandled. %v", msg.EventId, m)
```
Make this explicit so that if we add future websocket connections we'll know what this is referring to.
```suggestion
				log.Errorf("wsDcrdata message of type %v unhandled. %v",
					msg.EventId, m)
```

lukebp commented on politeiawww/www.go:
```diff
@@ -331,6 +331,14 @@ func _main() error {
 
+	// Get initial value for best block. This will subsequently be updated by
+	// the websocket connection to dcrdata.
+	bb, err := p.getBestBlockDecredPlugin()
```
I don't like this because if dcrdata is down then you won't be able to start up politeiawww.  It will error out here and exit.  The best block should be lazy loaded in `getBestBlock()`.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -90,9 +102,90 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+				// Reconnect
+				c, err := newDcrdataWSClient()
+				if err != nil {
+					goto wait
```
This could use a `log.Errorf`.  Forgot this in the snippet I posted.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -131,6 +133,11 @@ type politeiawww struct {
 	wsDcrdata *wsDcrdata
+
+	// The current best block is cached and updated using a websocket
+	// subscription to dcrdata
```
Add to this comment that the fallback is the dcrdata best block route if the websocket connection is not active.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -90,9 +102,90 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+	}
+	_, err := w.client.Subscribe(newBlockSub)
+	if err != nil {
+		return fmt.Errorf("failed to subscribe: %v", err)
```
Should include the subscription.  Can you fix this in the other subscribe and unsubscribe functions too?
```suggestion
		return fmt.Errorf("failed to subscribe to %v: %v", newBlockSub, err)
```

martonp commented:
@lukebp Yes, I've tested all the error paths. I've also made a change to reset the cached block number to zero before reconnecting, so the old stale value will not be used after reconnection.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -858,6 +866,78 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+// and making necessary changes to the required changes to the politeiawww
+// state so that the service continues to function properly during and after
+// the reconnection.
+func (p *politeiawww) resetPiDcrdataWSSubs() {
```
Why do you need this?  

`if bb == 0 || !p.wsDcrdata.isSubscribed(newBlockSub) {`

This already makes sure that the decred plugin command is used when there is no subscription.

martonp replied:
There might be a subscription, but if no messages have been received yet, the cached value will be incorrect.

lukebp replied:
If no messages have been received yet, won't bb be 0 and it will use the decred plugin command?

martonp replied:
Are you asking why I'm resetting the cached value to 0? 

This is because if we are disconnected from dcrdata and this value is not set to zero, then in the time after we reconnect and before we receive the first message, the value from the last message before we were disconnected will be returned to the user.

lukebp replied:
Oh I see.  In that case, can't this line in `getBestBlock()`

`if bb == 0 || !p.wsDcrdata.isSubscribed(newBlockSub) {`

just be 

`if bb == 0 {` 

Since bb will be 0 when either:
1. The subscription is inactive
Or
2. The subscription is active but has not received the best block yet

martonp replied:
Yep that works.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -83,16 +96,100 @@ func (w *wsDcrdata) unsubFromAddr(address string) error {
+				// Reconnect
+				c, err := newDcrdataWSClient()
+				if err != nil {
+					log.Errorf(err.Error())
```
You need to give your error messages context when they're being printed to the console.  Also, why are you calling the `Error()` method instead of just printing `err`?

martonp replied:
I was calling err.Error() because that returns a string. I've updated it now to give more context though.

martonp commented:
@lukebp , this is ready to be reviewed again. I've added the `Close()` and `isShutdown()` methods, and I've also added a `Recieve()` method to the wsDcrdata type.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -26,13 +30,46 @@ var (
+	shutdown      bool
 	client        *client.Client      // dcrdata websocket client
 	subscriptions map[string]struct{} // Client subscriptions
+	recvMsgChan   chan *client.ClientMessage
```
It seems like you made this unnecessarily complicated.  Why do you need this channel?

Why not just check isShutdown() at the beginning of any method that makes an external call (subToAddr, reconnect, etc) and return errShutdown if true.

martonp replied:
I added this to let listeners know that wsDcrdata is being shutdown so they can gracefully stop listening. Otherwise, the channel will output a null message and the listener will attempt to reconnect (and fail since isShutdown() is checked in the beginning of the reconnect method). Nothing bad will happen without this, other than a spew of error messages when there is actually no error.

lukebp replied:
Ok, I see what you're saying.  The correct way to do this would be to get rid of the channel and have a wsDcrdata method for all of the external calls that are being made (ping, subscribe).  Those methods will check isShutdown before making the call.  

You'd need to replace all of the direct calls with the wsDcrdata methods.  Ex: replace all the `w.client.Subscribe(event)` with `w.Subscribe(event)`.

Make sense?

martonp replied:
Yeah makes sense.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -26,13 +30,46 @@ var (
 	// event that has not yet been subscribed to.
 	errSubNotFound = errors.New("subscription not found")
+
+	errShutdown = errors.New("ws connection is shut down")
```
Needs a comment.

This is the second time I've had to leave this same comment in this PR (prev was for newBlockSub).  This is sloppy.  Step it up and review your code more carefully.

martonp replied:
Sorry about that I'll pay more attention.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -26,13 +30,46 @@ var (
+type wsDcrdataShutdown struct{}
+
+// Close closes the dcrdata websocket client and lets listeners know that this is happening.
+func (w *wsDcrdata) Close() error {
```
No need to export this when its in package main.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -51,8 +90,20 @@ func (w *wsDcrdata) removeSub(event string) {
+
 // isSubscribed returns whether the client is subscribed to the provided event.
 func (w *wsDcrdata) isSubscribed(event string) bool {
+	if w.isShutdown() {
```
You don't need to check this if the subscriptions are cleared out properly in the close() method.  You really will only need to check isShutdown before making external calls.  Like subscribe or ping.

chappjc commented on politeiawww/politeiawww.go:
```diff
@@ -858,6 +866,95 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+				if err != nil {
+					log.Errorf("Unable to reconnect to dcrdata, no longer " +
+						"listening.")
+					break listenerLoop
```
If dcrdata is down for a short period, I think you would want to continue to retry the reconnection.  What is the plan for connectivity issues that last more than an instant?

martonp replied:
Yes it does. You can see this in the reconnect method in wsdcrdata.go.

chappjc commented on politeiawww/politeiawww.go:
```diff
@@ -858,6 +866,95 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+				log.Debugf("wsDcrdata message WebsocketBlock(height=%s)",
+					m.Block.Height)
+				p.updateBestBlock(uint64(m.Block.Height))
+			case *pstypes.HangUp:
```
The current pubsub server will also be sending ping signals (`pstypes.SigPingAndUserCount`) at a regular interval.  Might as well catch that so it doesn't end up in the error logs.

chappjc replied:
To expand on this, the pings were supposed to be going out from the server -> clients, but there was a bug.  With the fix in, the client can/should set their read timeout appropriately (at least `pubsub.PingInterval`), here: https://github.com/decred/politeia/pull/1021/files#diff-ec12e4d6d15a45b1c30aa86a9f67b420R192.  Please have a look at the dcrdata PR that fixes the ping and exports the server ping interval: https://github.com/decred/dcrdata/pull/1645/files

If I understand the current read timeout strategy in this PR, it is to reconnect.  Ideally that would not be necessary as regular pings will be coming through so the client knows the connection is healthy.

chappjc replied:
Oh, sorry I forget that read timeouts in psclient don't raise to an error in the consumer.  Although a big change in https://github.com/decred/dcrdata/pull/1645/files is that legit read timeouts will be errors since the server sent ping should be relied upon for health.

I'd suggest updating the the pubsub dependency after I merge https://github.com/decred/dcrdata/pull/1645 (today) and set the read timeout to `pubsub.PingInterval + something` or use the default read timeout va nil Opts which sets it to `pubsub.PingInterval * 10 / 9`.

chappjc replied:
The server updates are merged and deployed.  Now you'd just need to update the pubsub import and set the read timeout appropriately.  e.g. https://github.com/decred/politeia/commit/53ee8215937fbbe6ed9db37bca92220867922619
You can cherry-pick 53ee8215937fbbe6ed9db37bca92220867922619 if it works for you.

chappjc replied:
You'll also want that empty `case *pstypes.SigPingAndUserCount:` too.

martonp replied:
@chappjc Sounds good, but could you also deploy to the testnet dcrdata? That one is still using 3.1.0.

chappjc replied:
Now deployed to testnet.dcrdata.org.

martonp replied:
Thank you sir.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -123,8 +214,97 @@ func newWSDcrdata() (*wsDcrdata, error) {
+		return errShutdown
+	}
+
+	done := make(chan struct{})
```
Looking at this again, does this channel actually do anything?  Wouldn't this function do the exact same thing if you took the go routine out?

martonp replied:
Yeah you're right, it's not needed.

martonp commented:
@lukebp This is ready to be reviewed again.

chappjc commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,100 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+				}
+				log.Infof("Successfully reconnected to dcrdata")
+			case int:
+				// Ping messages are of type int
```
Glad you found the right type for pings.  Sorry I said it was pstypes... type.  The type really should be made specific in a future psclient and pubsub/types.

martonp replied:
It's all good, but yes that would definitely be nicer.

lukebp commented on politeiawww/cmsaddresswatcher.go:
```diff
@@ -30,7 +30,7 @@ const (
 
 func (p *politeiawww) addWatchAddress(address string) {
-	err := p.wsDcrdata.subToAddr(address)
+	err := p.wsDcrdata.subscribe(addrSubPrefix + address)
```
Don't do this.  Use the `wsDcrdata.subscribe()` method in the `subToAddr` method instead of the client `Subscribe()` method.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,100 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+	}
+
+	go func() {
+	listenerLoop:
```
What's the purpose of this `listenerLoop` tag?  Will a simple `return` not suffice instead of a `break listenerLoop`?

martonp replied:
you're right it doesn't make sense anymore.. I think it made sense in a previous version.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,100 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+	go func() {
+	listenerLoop:
+		for {
+			msg, ok := <-p.wsDcrdata.client.Receive()
```
This should be done as a `wsDcrdata` method as well like the other client calls where it first checks if the ws connection is shutdown and returns an error if it is.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,100 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+
+				log.Errorf("Error receiving msg from dcrdata, msg: %v", msg)
+				err = p.resetPiDcrdataWSSubs()
+				if err != nil {
```
This needs to explicitly check for a `errShutdown`.  If it is a `errShutdown` then log something like `Dcrdata websocket connection closed` and return.

If there is an error that is not a `errShutdown` then log the actual error.  `resetPiDcrdataWSSubs` only returns `errShutdown` right now, but that may change in the future so its better to write it as if it does so that it doesn't get missed if other errors are ever added.

Same for the other resetPiDcrdataWSSubs below.

martonp replied:
Sounds good, I'll just log the error from wsDcrdata which says "dcrdata ws connection is shut down".

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -123,8 +177,86 @@ func newWSDcrdata() (*wsDcrdata, error) {
+// same subscriptions as the previous client.
+func (w *wsDcrdata) reconnect() error {
+	if w.isShutdown() {
+		log.Errorf("wsDcrdata is shutdown, not reconnecting")
```
Don't log this here.  The error is being returned so its up to the calling function on whether or not to log it.

martonp commented:
@lukebp This is ready to be reviewed again.

lukebp commented on politeiawww/wsdcrdata.go:
```diff
@@ -123,8 +193,88 @@ func newWSDcrdata() (*wsDcrdata, error) {
+		if len(prevSubscriptions) > 0 {
+			log.Debugf("wsDcrdata reconnect waiting %v", timeToWait)
+			time.Sleep(timeToWait)
+			timeToWait = 2 * timeToWait
```
Can we max out an an hour so it doesn't keep backing off forever?

```
			// Increase wait time until it reaches an hour then
			// try to reconnect once per hour.
			timeToWait = 2 * timeToWait
			if timeToWait > 60 {
				timeToWait = 60
			}
```

martonp replied:
You have to do time.Hour. This will make it 60 nanoseconds.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,104 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+	go func() {
+		for {
+			receiver, err := p.wsDcrdata.receive()
+			if err != nil {
```
Explicitly check for an errShutdown and exit gracefully.  Only log the error if its not an errShutdown.  Same for below.

```
			if err == errShutdown {
				log.Infof("Dcrdata websocket closed")
				return
			} else if err != nil {
				log.Errorf("wsDcrdata receive: %v", err)
				log.Infof("Dcrdata websocket closed")
				return
			}

```

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,104 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+				// After closing wsDcrdata, the channel channel will send a nil
+				// message. This check is just here to avoid a spew of
+				// unnecessary error messages.
+				if p.wsDcrdata.isShutdown() {
```
Is this necessary since its already going to be checking for an errShutdown on the receive method?

martonp replied:
Yes, it's needed because wsDcrdata can be shutdown after the receive method is called, but before the channel has received any value.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,104 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+					return
+				}
+
+				log.Errorf("Error receiving msg from dcrdata, msg: %v", msg)
```
I don't understand this, won't the message be nil here?  Add a comment explaining this scenario.

martonp replied:
Yeah you're right, this is pointless.

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,104 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+
+				log.Errorf("Error receiving msg from dcrdata, msg: %v", msg)
+				err = p.resetPiDcrdataWSSubs()
+				if err != nil {
```
Explicitly check for errShutdown.

```
				if err == errShutdown {
					log.Infof("Dcrdata websocket closed")
					return
				} else if err != nil {
					log.Errorf("resetPiDcrdataWSSub: %v", err)
					log.Infof("Dcrdata websocket closed")
					return
				}
```

lukebp commented on politeiawww/politeiawww.go:
```diff
@@ -859,6 +867,104 @@ func (p *politeiawww) handleWebsocketWrite(wc *wsContext) {
+			case *pstypes.HangUp:
+				log.Infof("Dcrdata has hung up. Will reconnect...")
+				err = p.resetPiDcrdataWSSubs()
+				if err != nil {
```
Explicitly check for errShutdown.

```
				if err == errShutdown {
					log.Infof("Dcrdata websocket closed")
					return
				} else if err != nil {
					log.Errorf("resetPiDcrdataWSSub: %v", err)
					log.Infof("Dcrdata websocket closed")
					return
				}
```

martonp commented:
@lukebp Ready for review.

martonp commented:
@lukebp It's fixed.
